### PR TITLE
provide endpoint for current project docs in redis (wip)

### PR DESCRIPTION
**Pull request by briangough**
add an endpoint GET /project/:project_id which returns all the docs currently in the docupdater for the given project

### CODE REVIEW

File: app.coffee [hunk 1]
```diff
@@ -38,6 +38,7 @@ app.param 'doc_id', (req, res, next, doc_id) ->
 		next new Error("invalid doc id")
 
 app.get    '/project/:project_id/doc/:doc_id',                          HttpController.getDoc
+app.get    '/project/:project_id',                                      HttpController.getProjectDocs
```
Comment (jpallen) [inline app.coffee]: `/project/:project_id/doc` would be more consistent with our routes for a 'list docs' end point.

File: app/coffee/HttpController.coffee [hunk 2]
```diff
@@ -37,6 +37,23 @@ module.exports = HttpController =
 			size += (line.length + 1)
 		return size
 
+	getProjectDocs: (req, res, next = (error) ->) ->
+		project_id = req.params.project_id
+		# filter is string of existing docs "id:version,id:version,..."
+		filterItems = req.query?.filter?.split(',') or []
```
Comment (jpallen) [inline app/coffee/HttpController.coffee]: Can we call this `ignore` or `exclude` instead. `filter` made me initially think it was the ones we wanted, no didn't want.

File: app/coffee/HttpController.coffee [hunk 3]
```diff
@@ -37,6 +37,23 @@ module.exports = HttpController =
...
+		timer = new Metrics.Timer("http.getAllDocs")
+		excludeVersions = {}
+		for item in filterItems
+			[id,version] = item?.split(':')
+			excludeVersions[id] = version
+		logger.log {project_id: project_id, excludeVersions: excludeVersions}, "excluding versions"
+		ProjectManager.getProjectDocs project_id, excludeVersions, (error, result) ->
+			timer.done()
+			return next(error) if error?
+			logger.log project_id: project_id, result: result, "got docs via http"
```
Comment (jpallen) [inline app/coffee/HttpController.coffee]: `result` is going to be a big log line here, that we'll hit a lot. Don't include it, or don't include the lines?

File: app.coffee (hunk 1 above)
Comment (briangough) [inline app.coffee]: changing that now

File: app/coffee/HttpController.coffee (hunk 2 above)
Comment (briangough) [inline app/coffee/HttpController.coffee]: will change it to exclude

File: app/coffee/HttpController.coffee (hunk 3 above)
Comment (briangough) [inline app/coffee/HttpController.coffee]: have trimmed the logs

File: app/coffee/RedisManager.coffee [hunk 4]
```diff
@@ -88,7 +88,23 @@ module.exports = RedisManager =
...
 			return callback(error) if error?
-			rclient.srem keys.docsInProject(project_id:project_id), doc_id, callback
+			multi = rclient.multi()
+			multi.srem keys.docsInProject(project_id:project_id), doc_id
+			if keys.projectState?
+				multi.del keys.projectState(project_id:project_id)
+			multi.exec callback
+
+	checkOrSetProjectState: (project_id, newState, callback = (error, stateChanged) ->) ->
+		if keys.projectState?
```
Comment (jpallen) [inline app/coffee/RedisManager.coffee]: Is there any reason that this would not be present?

File: app/coffee/HttpController.coffee [hunk 5]
```diff
@@ -37,6 +37,28 @@ module.exports = HttpController =
 			size += (line.length + 1)
 		return size
 
+	getProjectDocs: (req, res, next = (error) ->) ->
+		project_id = req.params.project_id
+		projectStateHash = req.query?.state
+		# exclude is string of existing docs "id:version,id:version,..."
+		excludeItems = req.query?.exclude?.split(',') or []
```
Comment (jpallen) [inline app/coffee/HttpController.coffee]: We're not actually sending this from web?

File: app/coffee/ProjectManager.coffee [hunk 6]
```diff
@@ -56,3 +57,35 @@ module.exports = ProjectManager =
...
+						jobs.push (cb) ->
+							# check the doc version first
+							RedisManager.getDocVersion doc_id, (error, version) ->
+								return cb(error) if error?
+								# skip getting the doc if we already have that version
+								return cb() if version is excludeVersions[doc_id]
+								# otherwise get the doc lines from redis
+								RedisManager.getDocLines doc_id, (error, lines) ->
+									return cb(error) if error?
+									docs.push {_id: doc_id, lines: lines, rev: version}
```
Comment (jpallen) [inline app/coffee/ProjectManager.coffee]: We use `v` for the ShareJS version, and `rev` for the Mongo version (everything time it's flushed to Mongo). `rev` is used as the version numbers for keeping Dropbox etc too. The doc updater only cares about `v` though, so it should return this as `v`.

File: app/coffee/RedisManager.coffee (hunk 4 above)
Comment (briangough) [inline app/coffee/RedisManager.coffee]: only if it wasn't in chef, but it's there now so I'll remove it

File: app/coffee/HttpController.coffee (hunk 5 above)
Comment (briangough) [inline app/coffee/HttpController.coffee]: no not currently, it's for future optimisation

File: app/coffee/ProjectManager.coffee (hunk 6 above)
Comment (briangough) [inline app/coffee/ProjectManager.coffee]: ok, will update

**Pull request merged**